File: scripts/validate-dotfiles.ml
#!/usr/bin/env ocaml

#directory "+unix"

#load "unix.cma"

(** * Dotfiles Repository Validation Script (OCaml Edition) * * A composable
    validation framework using a Rules API. * Each rule is a function that
    returns a validation result, * and rules can be easily composed together. *)

(* ========================================================================= *)
(* TYPES *)
(* ========================================================================= *)

type severity = Error | Warning | Info

type issue = {
  severity : severity;
  message : string;
  file : string option;
  fix_suggestion : string option;
}

type validation_result = {
  rule_name : string;
  passed : bool;
  issues : issue list;
}

type rule = unit -> validation_result
type config = { dotfiles_dir : string; verbose : bool; fix_mode : bool }

(* ========================================================================= *)
(* ANSI COLORS *)
(* ========================================================================= *)

module Color = struct
  let reset = "\x1b[0m"
  let bold = "\x1b[1m"
  let red = "\x1b[31m"
  let green = "\x1b[32m"
  let yellow = "\x1b[33m"
  let blue = "\x1b[34m"
  let cyan = "\x1b[36m"
  let success = green ^ "✓" ^ reset
  let failure = red ^ "✗" ^ reset
  let warning = yellow ^ "⚠" ^ reset
  let info = cyan ^ "ℹ" ^ reset
end

(* ========================================================================= *)
(* UTILITIES *)
(* ========================================================================= *)

let read_file path =
  try
    let ic = open_in path in
    let rec read_lines acc =
      try
        let line = input_line ic in
        read_lines (line :: acc)
      with End_of_file ->
        close_in ic;
        List.rev acc
    in
    Some (read_lines [])
  with _ -> None

let file_exists path =
  try
    let _ = Unix.stat path in
    true
  with Unix.Unix_error _ -> false

let read_command cmd =
  try
    let ic = Unix.open_process_in cmd in
    let rec read_lines acc =
      try
        let line = input_line ic in
        read_lines (line :: acc)
      with End_of_file ->
        let _ = Unix.close_process_in ic in
        List.rev acc
    in
    read_lines []
  with _ -> []

let is_tracked_by_git config path =
  let cmd =
    Printf.sprintf "cd %s && git ls-files --error-unmatch %s 2>/dev/null"
      config.dotfiles_dir path
  in
  match read_command cmd with [] -> false | _ -> true

let is_ignored_by_git config path =
  let cmd =
    Printf.sprintf "cd %s && git check-ignore %s 2>/dev/null"
      config.dotfiles_dir path
  in
  match read_command cmd with [] -> false | _ -> true

let get_tracked_files config =
  let cmd = Printf.sprintf "cd %s && git ls-files" config.dotfiles_dir in
  read_command cmd

(* ========================================================================= *)
(* TOML PARSING (Simple) *)
(* ========================================================================= *)

module TomlParser = struct
  type value = String of string | Section of (string * value) list

  let parse_line line =
    let trimmed = String.trim line in
    if trimmed = "" || String.get trimmed 0 = '#' then None
    else if String.get trimmed 0 = '[' then
      (* Section header *)
      let len = String.length trimmed in
      let section_name = String.sub trimmed 1 (len - 2) in
      Some (`Section section_name)
    else
      (* Key-value pair *)
      try
        let eq_pos = String.index trimmed '=' in
        let key = String.trim (String.sub trimmed 0 eq_pos) in
        let value =
          String.trim
            (String.sub trimmed (eq_pos + 1)
               (String.length trimmed - eq_pos - 1))
        in
        (* Remove quotes from key and value *)
        let clean_str s =
          let s = String.trim s in
          if
            String.length s >= 2
            && String.get s 0 = '"'
            && String.get s (String.length s - 1) = '"'
          then String.sub s 1 (String.length s - 2)
          else s
        in
        Some (`KeyValue (clean_str key, clean_str value))
      with Not_found -> None

  let parse_toml lines =
    let rec parse_sections acc current_section lines =
      match lines with
      | [] -> acc
      | line :: rest -> (
          match parse_line line with
          | None -> parse_sections acc current_section rest
          | Some (`Section name) -> parse_sections acc (Some name) rest
          | Some (`KeyValue (k, v)) -> (
              match current_section with
              | None -> parse_sections acc None rest
              | Some section ->
                  let new_acc = (section, k, v) :: acc in
                  parse_sections new_acc current_section rest))
    in
    parse_sections [] None lines

  let extract_files toml_data =
    List.filter_map
      (fun (section, key, value) ->
        if
          String.contains section '.'
          && (String.sub section (String.length section - 6) 6 = ".files"
             || String.sub section (String.length section - 5) 5 = ".files")
        then
          let group = String.sub section 0 (String.index section '.') in
          Some (key, value, group)
        else None)
      toml_data
end

(* ========================================================================= *)
(* VALIDATION RULES *)
(* ========================================================================= *)

module Rules = struct
  (* Rule: Dotter config files exist *)
  let dotter_configs_exist config : validation_result =
    let global_toml =
      Filename.concat
        (Filename.concat config.dotfiles_dir ".dotter")
        "global.toml"
    in

    let issues = [] in
    let issues =
      if not (file_exists global_toml) then
        {
          severity = Error;
          message = "Dotter global.toml not found";
          file = Some global_toml;
          fix_suggestion = None;
        }
        :: issues
      else issues
    in

    {
      rule_name = "Dotter configuration files exist";
      passed = issues = [];
      issues;
    }

  (* Rule: All files referenced in dotter config exist and are tracked *)
  let dotter_files_tracked config : validation_result =
    let global_toml =
      Filename.concat
        (Filename.concat config.dotfiles_dir ".dotter")
        "global.toml"
    in
    let macos_toml =
      Filename.concat
        (Filename.concat config.dotfiles_dir ".dotter")
        "macos.toml"
    in

    let parse_config path =
      match read_file path with
      | None -> []
      | Some lines ->
          let toml_data = TomlParser.parse_toml lines in
          TomlParser.extract_files toml_data
    in

    let global_files =
      if file_exists global_toml then parse_config global_toml else []
    in
    let macos_files =
      if file_exists macos_toml then parse_config macos_toml else []
    in
    let all_files = global_files @ macos_files in

    if config.verbose then
      Printf.printf "%sℹ Found %d files referenced in dotter configs%s\n"
        Color.cyan (List.length all_files) Color.reset;

    let check_file issues (source, _, group) =
      let filepath = Filename.concat config.dotfiles_dir source in
      if not (file_exists filepath) then
        {
          severity = Error;
          message = Printf.sprintf "File missing: %s (from %s)" source group;
          file = Some source;
          fix_suggestion = None;
        }
        :: issues
      else if not (is_tracked_by_git config source) then
        if is_ignored_by_git config source then
          {
            severity = Error;
            message =
              Printf.sprintf "File ignored by git: %s (from %s)" source group;
            file = Some source;
            fix_suggestion =
              Some (Printf.sprintf "Add to .gitignore: !%s" source);
          }
          :: issues
        else
          {
            severity = Warning;
            message =
              Printf.sprintf "File not tracked: %s (from %s)" source group;
            file = Some source;
            fix_suggestion = Some (Printf.sprintf "Run: git add %s" source);
          }
          :: issues
      else issues
    in

    let issues = List.fold_left check_file [] all_files in

    {
      rule_name = "Dotter files exist and are tracked";
      passed = List.for_all (fun i -> i.severity = Warning) issues;
      issues = List.rev issues;
    }

  (* Rule: No broken symlinks *)
  let no_broken_symlinks config : validation_result =
    let tracked = get_tracked_files config in
    let check_symlink file =
      let path = Filename.concat config.dotfiles_dir file in
      try
        let stats = Unix.lstat path in
        if stats.Unix.st_kind = Unix.S_LNK then
          try
            let _ = Unix.stat path in
            None
          with Unix.Unix_error _ ->
            Some
              {
                severity = Error;
                message = Printf.sprintf "Broken symlink: %s" file;
                file = Some file;
                fix_suggestion = None;
              }
        else None
      with Unix.Unix_error _ -> None
    in
    let issues = List.filter_map check_symlink tracked in
    { rule_name = "No broken symlinks"; passed = issues = []; issues }

  (* Rule: TOML files are valid *)
  let toml_files_valid config : validation_result =
    let tracked = get_tracked_files config in
    let toml_files =
      List.filter (fun f -> Filename.check_suffix f ".toml") tracked
    in

    let check_toml file =
      let path = Filename.concat config.dotfiles_dir file in
      match read_file path with
      | None ->
          Some
            {
              severity = Error;
              message = Printf.sprintf "Cannot read TOML file: %s" file;
              file = Some file;
              fix_suggestion = None;
            }
      | Some lines -> (
          (* Basic validation - just try to parse *)
          try
            let _ = TomlParser.parse_toml lines in
            None
          with e ->
            Some
              {
                severity = Error;
                message = Printf.sprintf "Invalid TOML syntax: %s" file;
                file = Some file;
                fix_suggestion = None;
              })
    in

    let issues = List.filter_map check_toml toml_files in
    {
      rule_name =
        Printf.sprintf "All %d TOML files are valid" (List.length toml_files);
      passed = issues = [];
      issues;
    }

end

(* ========================================================================= *)
(* RULE COMPOSITION *)
(* ========================================================================= *)

module Validator = struct
  let run_rule config rule =
    if config.verbose then
      Printf.printf "%sChecking: %s%s\n" Color.blue "..." Color.reset;
    rule ()

  let run_rules config rules = List.map (fun rule -> run_rule config rule) rules

  let print_result result =
    let status_icon = if result.passed then Color.success else Color.failure in
    Printf.printf "%s %s\n" status_icon result.rule_name;

    List.iter
      (fun issue ->
        let icon =
          match issue.severity with
          | Error -> Color.failure
          | Warning -> Color.warning
          | Info -> Color.info
        in
        let file_str =
          match issue.file with
          | Some f -> Printf.sprintf " (%s)" f
          | None -> ""
        in
        Printf.printf "  %s %s%s\n" icon issue.message file_str;

        match issue.fix_suggestion with
        | Some suggestion ->
            Printf.printf "    %s%s%s\n" Color.cyan suggestion Color.reset
        | None -> ())
      result.issues

  let summarize results config =
    Printf.printf "\n%s%s%s\n" Color.bold (String.make 60 '=') Color.reset;

    let total_issues =
      List.fold_left (fun acc r -> acc + List.length r.issues) 0 results
    in
    let errors =
      List.fold_left
        (fun acc r ->
          acc + List.length (List.filter (fun i -> i.severity = Error) r.issues))
        0 results
    in
    let warnings = total_issues - errors in

    if errors > 0 then begin
      Printf.printf
        "%s Validation failed: %d issue(s) found (%d errors, %d warnings)%s\n"
        Color.failure total_issues errors warnings Color.reset;

      if config.fix_mode then begin
        Printf.printf "\n%sFix suggestions:%s\n\n" Color.bold Color.reset;
        List.iter
          (fun result ->
            List.iter
              (fun issue ->
                match issue.fix_suggestion with
                | Some suggestion ->
                    Printf.printf "%s%s%s\n" Color.green suggestion Color.reset
                | None -> ())
              result.issues)
          results
      end;
      1
    end
    else if warnings > 0 then begin
      Printf.printf "%s Validation completed with %d warning(s)%s\n"
        Color.warning warnings Color.reset;
      0
    end
    else begin
      Printf.printf "%s All validations passed!%s\n" Color.success Color.reset;
      0
    end
end

(* ========================================================================= *)
(* MAIN *)
(* ========================================================================= *)

let () =
  let verbose = ref false in
  let fix_mode = ref false in
  let show_help = ref false in

  let args = Array.to_list Sys.argv in
  List.iter
    (fun arg ->
      match arg with
      | "-v" | "--verbose" -> verbose := true
      | "-f" | "--fix" -> fix_mode := true
      | "-h" | "--help" -> show_help := true
      | _ -> ())
    (List.tl args);

  if !show_help then begin
    Printf.printf "Usage: validate-dotfiles.ml [options]\n\n";
    Printf.printf "Options:\n";
    Printf.printf "  -f, --fix       Show fix suggestions\n";
    Printf.printf "  -v, --verbose   Show detailed output\n";
    Printf.printf "  -h, --help      Show this help message\n\n";
    Printf.printf "Exit codes:\n";
    Printf.printf "  0 - All validations passed\n";
    Printf.printf "  1 - Validation failures found\n";
    Printf.printf "  2 - Critical error\n";
    exit 0
  end;

  let dotfiles_dir =
    try Sys.getenv "DOTFILES_DIR"
    with Not_found ->
      let home = Sys.getenv "HOME" in
      Filename.concat home ".dotfiles"
  in

  let config =
    { dotfiles_dir; verbose = !verbose; fix_mode = !fix_mode }
  in

  Printf.printf "\n%sValidating dotfiles repository...%s\n\n" Color.bold
    Color.reset;

  (* Define all validation rules *)
  let rules =
    [
      (fun () -> Rules.dotter_configs_exist config);
      (fun () -> Rules.dotter_files_tracked config);
      (fun () -> Rules.no_broken_symlinks config);
      (fun () -> Rules.toml_files_valid config);
    ]
  in

  (* Run all rules and collect results *)
  let results = Validator.run_rules config rules in

  (* Print each result *)
  List.iter Validator.print_result results;

  (* Summarize and exit with appropriate code *)
  let exit_code = Validator.summarize results config in
  exit exit_code

File: scripts/validate-dotfiles.hs
#!/usr/bin/env runhaskell

{-# LANGUAGE RecordWildCards #-}

{-
  Dotfiles Repository Validation Script (Haskell Edition)

  A composable validation framework using a Rules API.
  Each rule is a function that returns a validation result,
  and rules can be easily composed together.
-}

import           Control.Exception  (SomeException, catch)
import           Control.Monad      (filterM, forM, forM_, unless, when)
import           Data.Char          (isSpace)
import           Data.List          (isInfixOf, isPrefixOf, isSuffixOf)
import           Data.Maybe         (fromMaybe, mapMaybe)
import           System.Directory   (doesFileExist, getModificationTime,
                                     listDirectory)
import           System.Environment (getArgs, lookupEnv)
import           System.Exit        (ExitCode (..), exitWith)
import           System.FilePath    (takeExtension, (</>))
import           System.Process     (readProcess, readProcessWithExitCode)

-- ========================================================================
-- TYPES
-- ========================================================================

data Severity = Error | Warning | Info
  deriving (Show, Eq)

data Issue = Issue
  { issueSeverity      :: Severity
  , issueMessage       :: String
  , issueFile          :: Maybe String
  , issueFixSuggestion :: Maybe String
  } deriving Show

data ValidationResult = ValidationResult
  { ruleName :: String
  , passed   :: Bool
  , issues   :: [Issue]
  } deriving Show

type Rule = IO ValidationResult

data Config = Config
  { dotfilesDir :: FilePath
  , verbose     :: Bool
  , fixMode     :: Bool
  } deriving Show

-- ========================================================================
-- ANSI COLORS
-- ========================================================================

data Color = Reset | Bold | Red | Green | Yellow | Blue | Cyan

colorCode :: Color -> String
colorCode Reset  = "\x1b[0m"
colorCode Bold   = "\x1b[1m"
colorCode Red    = "\x1b[31m"
colorCode Green  = "\x1b[32m"
colorCode Yellow = "\x1b[33m"
colorCode Blue   = "\x1b[34m"
colorCode Cyan   = "\x1b[36m"

withColor :: Color -> String -> String
withColor c s = colorCode c ++ s ++ colorCode Reset

-- ========================================================================
-- LOGGING HELPERS
-- ========================================================================

success :: String -> IO ()
success msg = putStrLn $ withColor Green ("✓ " ++ msg)

failure :: String -> IO ()
failure msg = putStrLn $ withColor Red ("✗ " ++ msg)

warning :: String -> IO ()
warning msg = putStrLn $ withColor Yellow ("⚠ " ++ msg)

info :: String -> IO ()
info msg = putStrLn $ withColor Cyan ("ℹ " ++ msg)

verboseLog :: Config -> String -> IO ()
verboseLog Config{..} msg = when verbose $ putStrLn $ withColor Blue ("  " ++ msg)

-- ========================================================================
-- UTILITIES
-- ========================================================================

trim :: String -> String
trim = f . f where f = reverse . dropWhile isSpace

runGitCommand :: Config -> String -> IO (Maybe String)
runGitCommand Config{..} cmd = do
  (exitCode, stdout, _) <- readProcessWithExitCode "sh" ["-c", "cd " ++ dotfilesDir ++ " && " ++ cmd] ""
  case exitCode of
    ExitSuccess -> return $ Just (trim stdout)
    _           -> return Nothing

isTrackedByGit :: Config -> FilePath -> IO Bool
isTrackedByGit config path = do
  result <- runGitCommand config $ "git ls-files --error-unmatch " ++ path ++ " 2>/dev/null"
  return $ maybe False (not . null) result

isIgnoredByGit :: Config -> FilePath -> IO Bool
isIgnoredByGit config path = do
  result <- runGitCommand config $ "git check-ignore " ++ path ++ " 2>/dev/null"
  return $ maybe False (not . null) result

getTrackedFiles :: Config -> IO [FilePath]
getTrackedFiles config = do
  result <- runGitCommand config "git ls-files"
  return $ maybe [] lines result

isBrokenSymlink :: FilePath -> IO Bool
isBrokenSymlink path = do
  exists <- doesFileExist path
  return (not exists)

-- ========================================================================
-- TOML PARSING (Simple)
-- ========================================================================

data TomlSection = TomlSection
  { sectionName    :: String
  , sectionEntries :: [(String, String)]
  } deriving Show

parseTomlLine :: String -> Maybe (Either String (String, String))
parseTomlLine line
  | null trimmed || "#" `isPrefixOf` trimmed = Nothing
  | "[" `isPrefixOf` trimmed && "]" `isSuffixOf` trimmed =
      -- Extract section name by dropping first '[' and last ']'
      case trimmed of
        ('[':rest) -> case reverse rest of
                        (']':revName) -> Just . Left $ reverse revName
                        _             -> Nothing
        _ -> Nothing
  | "=" `isInfixOf` trimmed =
      let (key, rest) = break (== '=') trimmed
          value = dropWhile (== '=') rest
      in Just . Right $ (cleanQuotes $ trim key, cleanQuotes $ trim value)
  | otherwise = Nothing
  where
    trimmed = trim line
    cleanQuotes s = case s of
      ('"':rest) -> case reverse rest of
                      ('"':revBody) -> reverse revBody
                      _             -> s
      _ -> s

parseToml :: FilePath -> IO [(String, String, String)]
parseToml path = do
  content <- readFile path
  let ls = lines content
      parsed = mapMaybe parseTomlLine ls
  return $ extractFiles parsed [] Nothing
  where
    extractFiles [] acc _ = acc
    extractFiles (Left section : rest) acc _ = extractFiles rest acc (Just section)
    extractFiles (Right (k, v) : rest) acc (Just sect)
      | ".files" `isSuffixOf` sect =
          let group = takeWhile (/= '.') sect
          in extractFiles rest ((k, v, group) : acc) (Just sect)
      | otherwise = extractFiles rest acc (Just sect)
    extractFiles (_ : rest) acc Nothing = extractFiles rest acc Nothing

-- ========================================================================
-- VALIDATION RULES
-- ========================================================================

-- Rule: Dotter configuration files exist
dotterConfigsExist :: Config -> Rule
dotterConfigsExist config@Config{..} = do
  let globalToml = dotfilesDir </> ".dotter" </> "global.toml"
  exists <- doesFileExist globalToml

  let issues = if exists then [] else
        [Issue Error "Dotter global.toml not found" (Just globalToml) Nothing]

  return $ ValidationResult "Dotter configuration files exist" (null issues) issues

-- Rule: All files referenced in dotter config exist and are tracked
dotterFilesTracked :: Config -> Rule
dotterFilesTracked config@Config{..} = do
  let globalToml = dotfilesDir </> ".dotter" </> "global.toml"
      macosToml  = dotfilesDir </> ".dotter" </> "macos.toml"

  globalFiles <- parseToml globalToml `catch` (\(_ :: SomeException) -> return [])
  macosExists <- doesFileExist macosToml
  macosFiles <- if macosExists
                then parseToml macosToml `catch` (\(_ :: SomeException) -> return [])
                else return []

  let allFiles = globalFiles ++ macosFiles

  when verbose $ info $ "Found " ++ show (length allFiles) ++ " files referenced in dotter configs"

  issues <- fmap concat $ forM allFiles $ \(source, _, group) -> do
    let filepath = dotfilesDir </> source
    exists <- doesFileExist filepath
    if not exists then
      return [Issue Error ("File missing: " ++ source ++ " (from " ++ group ++ ")")
                    (Just source) Nothing]
    else do
      tracked <- isTrackedByGit config source
      if not tracked then do
        ignored <- isIgnoredByGit config source
        if ignored then
          return [Issue Error ("File ignored by git: " ++ source ++ " (from " ++ group ++ ")")
                       (Just source) (Just $ "Add to .gitignore: !" ++ source)]
        else
          return [Issue Warning ("File not tracked: " ++ source ++ " (from " ++ group ++ ")")
                       (Just source) (Just $ "Run: git add " ++ source)]
      else return []

  let onlyErrors = filter (\i -> issueSeverity i == Error) issues
  return $ ValidationResult "Dotter files exist and are tracked" (null onlyErrors) issues

-- Rule: No broken symlinks
noBrokenSymlinks :: Config -> Rule
noBrokenSymlinks config@Config{..} = do
  tracked <- getTrackedFiles config
  issues <- fmap concat $ forM tracked $ \file -> do
    let path = dotfilesDir </> file
    broken <- isBrokenSymlink path `catch` (\(_ :: SomeException) -> return False)
    return $ if broken
      then [Issue Error ("Broken symlink: " ++ file) (Just file) Nothing]
      else []

  return $ ValidationResult "No broken symlinks" (null issues) issues

-- Rule: TOML files are valid
tomlFilesValid :: Config -> Rule
tomlFilesValid config@Config{..} = do
  tracked <- getTrackedFiles config
  let tomlFiles = filter (".toml" `isSuffixOf`) tracked

  issues <- fmap concat $ forM tomlFiles $ \file -> do
    let path = dotfilesDir </> file
    result <- (parseToml path >> return Nothing) `catch`
              (\(_ :: SomeException) -> return $ Just file)
    case result of
      Nothing -> return []
      Just f  -> return [Issue Error ("Invalid TOML syntax: " ++ f) (Just f) Nothing]

  return $ ValidationResult ("All " ++ show (length tomlFiles) ++ " TOML files are valid")
                           (null issues) issues

-- ========================================================================
-- RULE COMPOSITION & EXECUTION
-- ========================================================================

runRule :: Config -> Rule -> IO ValidationResult
runRule config rule = do
  verboseLog config "Checking..."
  rule

runRules :: Config -> [Rule] -> IO [ValidationResult]
runRules config = mapM (runRule config)

printResult :: ValidationResult -> IO ()
printResult ValidationResult{..} = do
  let statusIcon = if passed then withColor Green "✓" else withColor Red "✗"
  putStrLn $ statusIcon ++ " " ++ ruleName

  forM_ issues $ \Issue{..} -> do
    let icon = case issueSeverity of
                 Error   -> withColor Red "✗"
                 Warning -> withColor Yellow "⚠"
                 Info    -> withColor Cyan "ℹ"
        fileStr = maybe "" (\f -> " (" ++ f ++ ")") issueFile
    putStrLn $ "  " ++ icon ++ " " ++ issueMessage ++ fileStr
    case issueFixSuggestion of
      Just suggestion -> putStrLn $ "    " ++ withColor Cyan suggestion
      Nothing         -> return ()

summarize :: Config -> [ValidationResult] -> IO Int
summarize Config{..} results = do
  putStrLn $ "\n" ++ withColor Bold (replicate 60 '=')

  let totalIssues = sum $ map (length . issues) results
      errors = length $ filter (\i -> issueSeverity i == Error) $ concatMap issues results
      warnings = totalIssues - errors

  if errors > 0 then do
    putStrLn $ withColor Red $ "✗ Validation failed: " ++ show totalIssues ++
               " issue(s) found (" ++ show errors ++ " errors, " ++ show warnings ++ " warnings)"

    when fixMode $ do
      putStrLn $ "\n" ++ withColor Bold "🔧 Fix suggestions:" ++ "\n"

      let ignoredFiles = [f | Issue _ _ (Just f) (Just s) <- concatMap issues results,
                             ".gitignore" `isInfixOf` s]
      unless (null ignoredFiles) $ do
        putStrLn $ withColor Cyan "Add these lines to .gitignore:"
        forM_ ignoredFiles $ \f -> putStrLn $ withColor Green $ "  !" ++ f
        putStrLn ""

      let untrackedFiles = [f | Issue _ _ (Just f) (Just s) <- concatMap issues results,
                               "git add" `isInfixOf` s]
      unless (null untrackedFiles) $ do
        putStrLn $ withColor Cyan "Run this command to track files:"
        putStrLn $ withColor Green $ "  git add " ++ unwords untrackedFiles
        putStrLn ""

    return 1
  else if warnings > 0 then do
    putStrLn $ withColor Yellow $ "⚠ Validation completed with " ++ show warnings ++ " warning(s)"
    return 0
  else do
    putStrLn $ withColor Green "✓ All validations passed!\n"
    return 0

-- ========================================================================
-- MAIN
-- ========================================================================

parseArgs :: [String] -> (Bool, Bool, Bool)
parseArgs args = (hasArg "help", hasArg "verbose", hasArg "fix")
  where
    hasArg opt = any (\a -> a == "-" ++ take 1 opt || a == "--" ++ opt) args

main :: IO ()
main = do
  args <- getArgs
  let (showHelp, isVerbose, showFix) = parseArgs args

  when showHelp $ do
    putStrLn "\nUsage: validate-dotfiles.hs [options]\n"
    putStrLn "Options:"
    putStrLn "  -f, --fix       Show fix suggestions"
    putStrLn "  -v, --verbose   Show detailed output"
    putStrLn "  -h, --help      Show this help message\n"
    putStrLn "Exit codes:"
    putStrLn "  0 - All validations passed"
    putStrLn "  1 - Validation failures found"
    putStrLn "  2 - Critical error\n"
    exitWith ExitSuccess

  dotfilesDirEnv <- lookupEnv "DOTFILES_DIR"
  homeDir <- lookupEnv "HOME"
  let dotfilesPath = fromMaybe (fromMaybe "." homeDir ++ "/.dotfiles") dotfilesDirEnv

  let config = Config
        { dotfilesDir = dotfilesPath
        , verbose = isVerbose
        , fixMode = showFix
        }

  putStrLn $ "\n" ++ withColor Bold "Validating dotfiles repository..." ++ "\n"

  -- Define all validation rules
  let rules =
        [ dotterConfigsExist config
        , dotterFilesTracked config
        , noBrokenSymlinks config
        , tomlFilesValid config
        ]

  -- Run all rules and collect results
  results <- runRules config rules

  -- Print each result
  mapM_ printResult results

  -- Summarize and exit with appropriate code
  exitCode <- summarize config results
  exitWith $ if exitCode == 0 then ExitSuccess else ExitFailure exitCode
